odl/test/solvers/functional/functional_test.py

```diff
@@ -510,52 +496,67 @@ def test_multiplication_with_vector(space):
     assert all_almost_equal(y_other_times_func(x), expected_result,
                             places=places)
 
+# Fixtures for test_functional_quadratic_perturb
+linear_term = simple_fixture('linear_term', [False, True])
+quadratic_term = simple_fixture('quadratic_term', [0.0, 2.13])
+
 
-def test_functional_linear_perturb(space):
-    """Test for the functional f(.) + <y, .>."""
+def test_functional_quadratic_perturb(space, linear_term, quadratic_term):
+    """Test for the functional f(.) + a | . |^2 + <y, .>."""
     # Less strict checking for single precision
     places = 3 if space.dtype == np.float32 else 5
 
-    # The translation; an element in the domain
-    linear_term = noise_element(space)
-
-    # Creating the functional ||x||_2^2 and add the linear perturbation
     orig_func = odl.solvers.L2NormSquared(space)
-    functional = odl.solvers.FunctionalLinearPerturb(orig_func, linear_term)
+
+    if linear_term:
+        linear_term_arg = None
+        linear_term = space.zero()
+    else:
+        linear_term_arg = linear_term = noise_element(space)
+
+    # Creating the functional ||x||_2^2 and add the quadratic perturbation
+    functional = odl.solvers.FunctionalQuadraticPerturb(
+        orig_func, quadratic_term=quadratic_term, linear_term=linear_term_arg)
 
     # Create an element in the space, in which to evaluate
     x = noise_element(space)
 
     # Test for evaluation of the functional
-    assert all_almost_equal(functional(x), x.norm()**2 + x.inner(linear_term),
+    assert all_almost_equal(functional(x),
+                            (orig_func(x) +
+                             quadratic_term * x.inner(x) +
+                             x.inner(linear_term)),
```

Comment:
I'm honestly not so sure about these kinds of tests. What I mean by "these tests" is tests of the *implementation* of something rather than the *intended functionality* of something.

What I think one should do instead (and this is not easy while implementing stuff) is to take a step back, ask oneself "What is this function supposed to do?" and then test that. Or test things that implicitly follow but aren't directly implemented.
In this case it's almost the same, but there are still some differences, for instance what happens with weighted spaces. Errors there won't be caught by this test.

Here I think it's better to test an explicit example, and when we feel we need to settle on some behavior regarding weighting (it will influence the gradient) we add a test for that.